// @carbon/ibmdotcom-react/lib/components/HorizontalRule/HorizontalRule.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _classnames2 = _interopRequireDefault(require("classnames"));

var _settings = _interopRequireDefault(require("@carbon/ibmdotcom-utilities/lib/utilities/settings/settings"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _settings2 = _interopRequireDefault(require("carbon-components/umd/globals/js/settings"));

/**
 * Copyright IBM Corp. 2016, 2020
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
var stablePrefix = _settings.default.stablePrefix;
var prefix = _settings2.default.prefix;
/**
 * Horizontal Rule component.
 */

var HorizontalRule = function HorizontalRule(_ref) {
  var _classnames;

  var style = _ref.style,
      size = _ref.size,
      contrast = _ref.contrast,
      weight = _ref.weight;
  return _react.default.createElement("hr", {
    "data-autoid": "".concat(stablePrefix, "--hr"),
    className: (0, _classnames2.default)("".concat(prefix, "--hr"), (_classnames = {}, (0, _defineProperty2.default)(_classnames, "".concat(prefix, "--hr--").concat(style), style), (0, _defineProperty2.default)(_classnames, "".concat(prefix, "--hr--").concat(contrast), contrast), (0, _defineProperty2.default)(_classnames, "".concat(prefix, "--hr--").concat(size), size), (0, _defineProperty2.default)(_classnames, "".concat(prefix, "--hr--").concat(weight), weight), _classnames))
  });
};

HorizontalRule.propTypes = {
  /**
   * Style of the HR. Choose from:
   *
   * | Name     | Description                                                                    |
   * | -------- | ------------------------------------------------------------------------------ |
   * | `solid`  | Default style variant of the horizontal rule                                   |
   * | `dashed` | Dashed/dotted style variant (style is applied in conjuction with `fluid` size) |
   */
  style: _propTypes.default.oneOf(['solid', 'dashed']),

  /**
   * Length of the HR. Choose from:
   *
   * | Name     | Description                                                         |
   * | -------- | ------------------------------------------------------------------- |
   * | `fluid`  | Default size variant - horizontal rule takes full width of the grid |
   * | `small`  | Shorter fixed-length variant with a max-width of 32px               |
   * | `medium` | Medium fixed-length variant with a max-width of 48px                |
   * | `large`  | Longer fixed-length variant with a max-width of 64px                |
   */
  size: _propTypes.default.oneOf(['fluid', 'small', 'medium', 'large']),

  /**
   * Contrast of the HR. Choose from:
   *
   * | Name              | Description                    |
   * | ----------------- | ------------------------------ |
   * | `medium-contrast` | Default contrast color variant |
   * | `low-contrast`    | Lighter contrast color variant |
   * | `high-contrast`   | Darker contrast color variant  |
   */
  contrast: _propTypes.default.oneOf(['medium-contrast', 'low-contrast', 'high-contrast']),

  /**
   * Weight of the HR. Choose from:
   *
   * | Name    | Description                                                                          |
   * | ------- | ------------------------------------------------------------------------------------ |
   * | `thin`  | Default weight variant                                                               |
   * | `thick` | Slightly thicker weight variant (this only applied in conjunction with `fluid` size) |
   */
  weight: _propTypes.default.oneOf(['thin', 'thick'])
};
var _default = HorizontalRule;
exports.default = _default;
