// @carbon/ibmdotcom-react/lib/internal/FeatureFlags.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DDS_CONTENTBLOCK_HEADLINES = exports.DDS_CALLOUT_DATA = exports.DDS_LOGO_GRID = exports.DDS_SIMPLE_OVERVIEW = exports.DDS_SIMPLEBENEFITS = exports.DDS_LANGUAGE_SELECTOR = exports.DDS_MASTHEAD_L1 = exports.DDS_FLAGS_ALL = void 0;

/**
 * Copyright IBM Corp. 2016, 2020
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

/**
 * This file contains the list of the default values of compile-time feature flags.
 */

/**
 * This flag will determine if all feature flags should be enabled
 *
 * @type {boolean}
 */
var DDS_FLAGS_ALL = process.env.DDS_FLAGS_ALL === 'true' || false;
/**
 * Feature flag to turn on the Masthead L1
 *
 * @type {boolean}
 */

exports.DDS_FLAGS_ALL = DDS_FLAGS_ALL;
var DDS_MASTHEAD_L1 = process.env.DDS_MASTHEAD_L1 === 'true' || DDS_FLAGS_ALL || false;
/**
 * Feature flag for the optional language selector in the footer
 *
 * @type {boolean}
 */

exports.DDS_MASTHEAD_L1 = DDS_MASTHEAD_L1;
var DDS_LANGUAGE_SELECTOR = process.env.DDS_LANGUAGE_SELECTOR === 'true' || DDS_FLAGS_ALL || false;
/**
 * This determines if the simplebenefits will be rendered or not
 *
 * @type {string | boolean}
 */

exports.DDS_LANGUAGE_SELECTOR = DDS_LANGUAGE_SELECTOR;
var DDS_SIMPLEBENEFITS = process.env.DDS_SIMPLEBENEFITS === 'true' || DDS_FLAGS_ALL || false;
/**
 * This determines if the simple overview will be rendered or not
 *
 * @type {string | boolean}
 */

exports.DDS_SIMPLEBENEFITS = DDS_SIMPLEBENEFITS;
var DDS_SIMPLE_OVERVIEW = process.env.DDS_SIMPLE_OVERVIEW === 'true' || DDS_FLAGS_ALL || false;
/**
 * This determines if the logo grid will be rendered or not
 *
 * @type {string | boolean}
 */

exports.DDS_SIMPLE_OVERVIEW = DDS_SIMPLE_OVERVIEW;
var DDS_LOGO_GRID = process.env.DDS_LOGO_GRID === 'true' || DDS_FLAGS_ALL || false;
/**
 * This determines if the logo grid will be rendered or not
 *
 * @type {string | boolean}
 */

exports.DDS_LOGO_GRID = DDS_LOGO_GRID;
var DDS_CALLOUT_DATA = process.env.DDS_CALLOUT_DATA === 'true' || DDS_FLAGS_ALL || false;
/**
 * This determines if content block - headlines will be rendered or not
 *
 * @type {string | boolean}
 */

exports.DDS_CALLOUT_DATA = DDS_CALLOUT_DATA;
var DDS_CONTENTBLOCK_HEADLINES = process.env.DDS_CONTENTBLOCK_HEADLINES === 'true' || DDS_FLAGS_ALL || false;
exports.DDS_CONTENTBLOCK_HEADLINES = DDS_CONTENTBLOCK_HEADLINES;
