// @carbon/ibmdotcom-react/lib/components/Masthead/MastheadProfile.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _OverflowMenu = _interopRequireDefault(require("../../internal/vendor/carbon-components-react/components/OverflowMenu/OverflowMenu"));

var _OverflowMenuItem = _interopRequireDefault(require("../../internal/vendor/carbon-components-react/components/OverflowMenuItem/OverflowMenuItem"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _settings = _interopRequireDefault(require("carbon-components/umd/globals/js/settings"));

/**
 * Copyright IBM Corp. 2016, 2020
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
var prefix = _settings.default.prefix;
/**
 * MastHead Profile component.
 */

var MastheadProfile = function MastheadProfile(_ref) {
  var overflowMenuProps = _ref.overflowMenuProps,
      overflowMenuItemProps = _ref.overflowMenuItemProps,
      profileMenu = _ref.profileMenu;

  /**
   * Masthead profile menu
   *
   * @returns {*} Masthead profile menu
   */
  var profileNav = profileMenu.map(function (item, i) {
    return _react.default.createElement(_OverflowMenuItem.default, (0, _extends2.default)({}, overflowMenuItemProps, {
      itemText: item.title,
      href: item.url,
      hasDivider: i > 0,
      key: i,
      primaryFocus: true
    }));
  });
  return _react.default.createElement(_OverflowMenu.default, (0, _extends2.default)({
    className: "".concat(prefix, "--header__action")
  }, overflowMenuProps), profileNav);
};

MastheadProfile.propTypes = {
  /**
   * The props for Carbon `<OverflowMenu>`.
   */
  overflowMenuProps: _propTypes.default.object,

  /**
   * The props for Carbon `<OverflowMenuItem>`.
   */
  overflowMenuItemProps: _propTypes.default.object,

  /**
   * The data to generate the Carbon `<OverflowMenuItem>`s from.
   */
  profileMenu: _propTypes.default.arrayOf(_propTypes.default.shape({
    title: _propTypes.default.string,
    url: _propTypes.default.string
  }))
};
var _default = MastheadProfile;
exports.default = _default;
