// @carbon/ibmdotcom-react/lib/components/Masthead/MastheadTopNav.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _react = _interopRequireWildcard(require("react"));

var _classnames3 = _interopRequireDefault(require("classnames"));

var _settings = _interopRequireDefault(require("@carbon/ibmdotcom-utilities/lib/utilities/settings/settings"));

var _HeaderMenu = _interopRequireDefault(require("../carbon-components-react/UIShell/HeaderMenu"));

var _HeaderMenuItem = _interopRequireDefault(require("../../internal/vendor/carbon-components-react/components/UIShell/HeaderMenuItem"));

var _HeaderName = _interopRequireDefault(require("../../internal/vendor/carbon-components-react/components/UIShell/HeaderName"));

var _HeaderNavigation = _interopRequireDefault(require("../../internal/vendor/carbon-components-react/components/UIShell/HeaderNavigation"));

var _MegaMenu = _interopRequireDefault(require("./MastheadMegaMenu/MegaMenu"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _settings2 = _interopRequireDefault(require("carbon-components/umd/globals/js/settings"));

/**
 * Copyright IBM Corp. 2016, 2020
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
var stablePrefix = _settings.default.stablePrefix;
var prefix = _settings2.default.prefix;
/**
 * Masthead top nav component.
 */

var MastheadTopNav = function MastheadTopNav(_ref) {
  var navigation = _ref.navigation,
      topNavProps = (0, _objectWithoutProperties2.default)(_ref, ["navigation"]);

  var _useState = (0, _react.useState)(false),
      _useState2 = (0, _slicedToArray2.default)(_useState, 2),
      overlay = _useState2[0],
      setOverlay = _useState2[1];
  /**
   * Top masthead navigation
   *
   * @returns {*} Top masthead navigation
   */


  var mastheadLinks = navigation.map(function (link, i) {
    var autoid = "".concat(stablePrefix, "--masthead-").concat(topNavProps.navType, "__l0-nav").concat(i);

    if (link.hasMenu || link.hasMegapanel) {
      return _react.default.createElement(_HeaderMenu.default, {
        "aria-label": link.title,
        menuLinkName: link.title,
        className: (0, _classnames3.default)((0, _defineProperty2.default)({}, "".concat(prefix, "--masthead__megamenu__l0-nav"), link.hasMegapanel)),
        autoId: autoid,
        key: i,
        setOverlay: setOverlay
      }, renderNav(link, autoid));
    } else {
      return _react.default.createElement(_HeaderMenuItem.default, {
        href: link.url,
        "data-autoid": autoid,
        key: i
      }, link.title);
    }
  });
  return _react.default.createElement(_react.default.Fragment, null, _react.default.createElement("div", {
    className: "".concat(prefix, "--header__nav-container")
  }, topNavProps.platform && _react.default.createElement(_HeaderName.default, {
    prefix: "",
    href: topNavProps.platform.url,
    "data-autoid": "".concat(stablePrefix, "--masthead-").concat(topNavProps.navType, "__l0-ecosystemname")
  }, topNavProps.platform.name), _react.default.createElement(_HeaderNavigation.default, {
    "aria-label": "IBM",
    "data-autoid": "".concat(stablePrefix, "--masthead__l0-nav")
  }, mastheadLinks)), _react.default.createElement("div", {
    className: (0, _classnames3.default)("".concat(prefix, "--masthead__overlay"), (0, _defineProperty2.default)({}, "".concat(prefix, "--masthead__overlay-show"), overlay))
  }));
};
/**
 * Loops through and renders a list of links for the masthead nav
 *
 * @param {object} link A list of links to be rendered
 * @param {string} autoid autoid predecessor for megamenu items/menu items data-autoids
 * @returns {object} JSX object
 */


function renderNav(link, autoid) {
  var navItems = [];

  if (link.hasMegapanel) {
    navItems.push(_react.default.createElement(_MegaMenu.default, {
      key: link.title,
      data: link,
      autoid: autoid
    }));
  } else {
    link.menuSections.forEach(function (section, i) {
      section.menuItems.forEach(function (item, j) {
        navItems.push(_react.default.createElement(_HeaderMenuItem.default, {
          href: item.url,
          "data-autoid": "".concat(autoid, "--subnav-col").concat(i, "-item").concat(j),
          key: item.title
        }, item.title));
      });
    });
  }

  return navItems;
}

MastheadTopNav.propTypes = {
  /**
   * Object containing top navigation elements.
   */
  navigation: _propTypes.default.arrayOf(_propTypes.default.shape({
    hasMenu: _propTypes.default.bool,
    title: _propTypes.default.string,
    url: _propTypes.default.string,
    menuSections: _propTypes.default.arrayOf(_propTypes.default.shape({
      menuItems: _propTypes.default.arrayOf(_propTypes.default.shape({
        title: _propTypes.default.string,
        url: _propTypes.default.string
      }))
    }))
  }))
};
var _default = MastheadTopNav;
exports.default = _default;
