// @carbon/ibmdotcom-react/lib/components/carbon-components-react/UIShell/HeaderMenu.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends3 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _classnames = _interopRequireDefault(require("classnames"));

var _ = _interopRequireDefault(require("@carbon/icons-react/lib/chevron--down/20"));

var _settings = _interopRequireDefault(require("carbon-components/umd/globals/js/settings"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _keyboard = require("../../../internal/vendor/carbon-components-react/internal/keyboard");

var _AriaPropTypes = require("../../../internal/vendor/carbon-components-react/prop-types/AriaPropTypes");

var _windowOrGlobal = _interopRequireDefault(require("window-or-global"));

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

var prefix = _settings.default.prefix;

var defaultRenderMenuContent = function defaultRenderMenuContent() {
  return _react.default.createElement(_.default, {
    className: "".concat(prefix, "--header__menu-arrow")
  });
};
/**
 * `HeaderMenu` is used to render submenu's in the `Header`. Most often children
 * will be a `HeaderMenuItem`. It handles certain keyboard events to help
 * with managing focus. It also passes along refs to each child so that it can
 * help manage focus state of its children.
 */


var HeaderMenu = /*#__PURE__*/function (_React$Component) {
  (0, _inherits2.default)(HeaderMenu, _React$Component);

  function HeaderMenu(props) {
    var _this;

    (0, _classCallCheck2.default)(this, HeaderMenu);
    _this = (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(HeaderMenu).call(this, props));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleOnClick", function (event) {
      _this.menuLinkRef.current.focus();

      _this.setState(function (prevState) {
        if (prevState.expanded) {
          var _root$document, _root$document$body;

          _this.props.setOverlay(false);

          (_root$document = _windowOrGlobal.default.document) === null || _root$document === void 0 ? void 0 : (_root$document$body = _root$document.body) === null || _root$document$body === void 0 ? void 0 : _root$document$body.classList.remove("".concat(prefix, "--body__lock-scroll"));
        } else {
          var _root$document2, _root$document2$body;

          _this.props.setOverlay(true);

          (_root$document2 = _windowOrGlobal.default.document) === null || _root$document2 === void 0 ? void 0 : (_root$document2$body = _root$document2.body) === null || _root$document2$body === void 0 ? void 0 : _root$document2$body.classList.add("".concat(prefix, "--body__lock-scroll"));
        }

        return {
          expanded: !prevState.expanded
        };
      });
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleOnKeyDown", function (event) {
      // Handle enter or space key for toggling the expanded state of the menu.
      if ((0, _keyboard.matches)(event, [_keyboard.keys.Enter, _keyboard.keys.Space])) {
        event.stopPropagation();
        event.preventDefault();

        _this.handleOnClick();

        return;
      }
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "checkMenuItems", function (event) {
      var megamenuItems = ["".concat(prefix, "--masthead__megamenu__category-headline"), "".concat(prefix, "--masthead__megamenu__category-group"), "".concat(prefix, "--masthead__megamenu__view-all-cta"), "".concat(prefix, "--masthead__megamenu__l0-nav"), "".concat(prefix, "--header__menu")];
      return megamenuItems.filter(function (item) {
        var _event$relatedTarget$;

        return (_event$relatedTarget$ = event.relatedTarget.parentElement.className) === null || _event$relatedTarget$ === void 0 ? void 0 : _event$relatedTarget$.includes(item);
      });
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleOnBlur", function (event) {
      if (!event.currentTarget.contains(event.relatedTarget)) {
        var _root$document3, _root$document3$body;

        _this.setState({
          expanded: false,
          selectedIndex: null
        });

        (_root$document3 = _windowOrGlobal.default.document) === null || _root$document3 === void 0 ? void 0 : (_root$document3$body = _root$document3.body) === null || _root$document3$body === void 0 ? void 0 : _root$document3$body.classList.remove("".concat(prefix, "--body__lock-scroll"));
      }

      if (!event.relatedTarget || !_this.checkMenuItems(event).length) {
        _this.props.setOverlay(false);
      }
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleMenuButtonRef", function (node) {
      if (_this.props.focusRef) {
        _this.props.focusRef(node);
      }

      _this.menuButtonRef = node;
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleItemRef", function (index) {
      return function (node) {
        _this.items[index] = node;
      };
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleMenuClose", function (event) {
      // Handle ESC keydown for closing the expanded menu.
      if ((0, _keyboard.matches)(event, [_keyboard.keys.Escape]) && _this.state.expanded) {
        event.stopPropagation();
        event.preventDefault();

        _this.setState(function () {
          return {
            expanded: false,
            selectedIndex: null
          };
        }); // remove overlay


        _this.props.setOverlay(false); // Return focus to menu button when the user hits ESC.


        _this.menuButtonRef.focus();

        return;
      }
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "_renderMenuItem", function (item, index) {
      return _react.default.cloneElement(item, {
        ref: _this.handleItemRef(index),
        role: 'none'
      });
    });
    _this.state = {
      // Used to manage the expansion state of the menu
      expanded: false,
      // Refers to the menuitem that is currently focused
      // Note: children should have `role="menuitem"` on node consuming ref
      selectedIndex: null
    };
    _this.items = [];
    _this.menuLinkRef = _react.default.createRef();
    return _this;
  }
  /**
   * Toggle the expanded state of the menu on click.
   */


  (0, _createClass2.default)(HeaderMenu, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          ariaLabel = _this$props['aria-label'],
          ariaLabelledBy = _this$props['aria-labelledby'],
          customClassName = _this$props.className,
          children = _this$props.children,
          MenuContent = _this$props.renderMenuContent,
          menuLinkName = _this$props.menuLinkName,
          autoId = _this$props.autoId;
      var accessibilityLabel = {
        'aria-label': ariaLabel,
        'aria-labelledby': ariaLabelledBy
      };
      var className = (0, _classnames.default)("".concat(prefix, "--header__submenu"), customClassName); // Notes on eslint comments and based on the examples in:
      // https://www.w3.org/TR/wai-aria-practices/examples/menubar/menubar-1/menubar-1.html#
      // - The focus is handled by the <a> menuitem, onMouseOver is for mouse
      // users
      // - aria-haspopup can definitely have the value "menu"
      // - aria-expanded is on their example node with role="menuitem"
      // - href can be set to javascript:void(0), ideally this will be a button

      return _react.default.createElement("li", {
        // eslint-disable-line jsx-a11y/mouse-events-have-key-events,jsx-a11y/no-noninteractive-element-interactions
        className: className,
        "data-autoid": autoId,
        onKeyDown: this.handleMenuClose,
        onBlur: this.handleOnBlur
      }, _react.default.createElement("a", (0, _extends3.default)((0, _defineProperty2.default)({
        // eslint-disable-line jsx-a11y/role-supports-aria-props,jsx-a11y/anchor-is-valid
        "aria-haspopup": "menu" // eslint-disable-line jsx-a11y/aria-proptypes
        ,
        "aria-expanded": this.state.expanded,
        className: "".concat(prefix, "--header__menu-item ").concat(prefix, "--header__menu-title"),
        href: "#",
        onClick: this.handleOnClick,
        onKeyDown: this.handleOnKeyDown,
        ref: this.handleMenuButtonRef,
        role: "menuitem",
        tabIndex: 0
      }, "ref", this.menuLinkRef), accessibilityLabel), menuLinkName, _react.default.createElement(MenuContent, null)), _react.default.createElement("ul", (0, _extends3.default)({}, accessibilityLabel, {
        className: "".concat(prefix, "--header__menu"),
        role: "menu"
      }), _react.default.Children.map(children, this._renderMenuItem)));
    }
    /**
     * Render an individual menuitem, passing along `role: 'none'` because the
     * host node <li> doesn't apply when in a <ul> with `role="menu"` and so we
     * need to revert the semantics.
     *
     * We also capture the `ref` for each child inside of `this.items` to properly
     * manage focus. In addition to this focus management, all items receive a
     * `tabIndex: -1` so the user won't hit a large number of items in their tab
     * sequence when they might not want to go through all the items.
     */

  }]);
  return HeaderMenu;
}(_react.default.Component);

(0, _defineProperty2.default)(HeaderMenu, "propTypes", _objectSpread({}, _AriaPropTypes.AriaLabelPropType, {
  /**
   * Provide a custom ref handler for the menu button
   */
  focusRef: _propTypes.default.func,

  /**
   * Optionally provide a tabIndex for the underlying menu button
   */
  tabIndex: _propTypes.default.number,

  /**
   * Provide a label for the link text
   */
  menuLinkName: _propTypes.default.string.isRequired,

  /**
   * Optional component to render instead of string
   */
  renderMenuContent: _propTypes.default.func,

  /**
   * function to toogle overlay that appears when opening menu
   */
  setOverlay: _propTypes.default.func
}));
(0, _defineProperty2.default)(HeaderMenu, "defaultProps", {
  renderMenuContent: defaultRenderMenuContent,
  setOverlay: function setOverlay() {}
});

var _default = _react.default.forwardRef(function (props, ref) {
  return _react.default.createElement(HeaderMenu, (0, _extends3.default)({}, props, {
    focusRef: ref
  }));
});

exports.default = _default;
