// @carbon/ibmdotcom-react/lib/components/carbon-components-react/UIShell/SideNavMenuWithBackForward.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _ = _interopRequireDefault(require("@carbon/icons-react/lib/arrow--right/20"));

var _2 = _interopRequireDefault(require("@carbon/icons-react/lib/chevron--left/20"));

var _settings = _interopRequireDefault(require("@carbon/ibmdotcom-utilities/lib/utilities/settings/settings"));

var _settings2 = _interopRequireDefault(require("carbon-components/umd/globals/js/settings"));

var _SideNavLink = _interopRequireDefault(require("../../../internal/vendor/carbon-components-react/components/UIShell/SideNavLink"));

var _SideNavMenu = _interopRequireDefault(require("./SideNavMenu"));

var _SideNavMenuItem = _interopRequireDefault(require("../../../internal/vendor/carbon-components-react/components/UIShell/SideNavMenuItem"));

var _classnames = _interopRequireDefault(require("classnames"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireWildcard(require("react"));

/**
 * Copyright IBM Corp. 2016, 2020
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
var stablePrefix = _settings.default.stablePrefix;
var prefix = _settings2.default.prefix;
/**
 * Side nav menu with back button
 *
 * @returns {*} SideNavMenuWithBackFoward menu items
 */

var SideNavMenuWithBackForward = function SideNavMenuWithBackForward(_ref) {
  var title = _ref.title,
      titleUrl = _ref.titleUrl,
      backButtonText = _ref.backButtonText,
      children = _ref.children,
      rest = (0, _objectWithoutProperties2.default)(_ref, ["title", "titleUrl", "backButtonText", "children"]);
  var refSideNavMenu = (0, _react.useRef)(null);
  var handleToggle = (0, _react.useCallback)(function (event, detail) {
    var sideNavMenuNode = refSideNavMenu.current;
    var sideNav = sideNavMenuNode.closest('.bx--side-nav');

    if (sideNav) {
      var list = Array.prototype.forEach.call(sideNav.querySelectorAll('.bx--side-nav__menu'), function (elem) {
        var hasExpandedSubmenu = elem.querySelector('.bx--side-nav__submenu[aria-expanded="true"]');
        elem.classList.toggle('bx--side-nav__menu--hasactivechildren', hasExpandedSubmenu);
      });
    }
  }, []);
  return _react.default.createElement(_SideNavMenu.default, {
    autoid: rest.autoid,
    title: title,
    onToggle: handleToggle,
    ref: refSideNavMenu
  }, _react.default.createElement(_SideNavMenuItem.default, {
    onClick: function onClick(event) {
      return event.preventDefault();
    },
    className: "".concat(prefix, "--masthead__side-nav--submemu-back"),
    "data-autoid": "".concat(stablePrefix, "--masthead-").concat(rest.navType, "-sidenav__l0-back"),
    isbackbutton: "true",
    role: "button",
    tabIndex: "0"
  }, _react.default.createElement(_2.default, null), backButtonText), titleUrl ? _react.default.createElement(_SideNavLink.default, {
    className: "".concat(prefix, "--masthead__side-nav--submemu-section-title"),
    href: titleUrl
  }, title, _react.default.createElement("div", {
    className: "".concat(prefix, "--masthead__side-nav--submemu-section-title__icon")
  }, _react.default.createElement(_.default, null))) : _react.default.createElement("li", {
    className: "".concat(prefix, "--masthead__side-nav--submemu-title")
  }, title), children);
};

SideNavMenuWithBackForward.propTypes = {
  /**
   * Submenu nav section title
   */
  title: _propTypes.default.string.isRequired,

  /**
   * Submenu nav section title url
   */
  titleUrl: _propTypes.default.string,

  /**
   * Back button text
   */
  backButtonText: _propTypes.default.string,

  /**
   * Sidenav child elements to be rendered
   */
  children: _propTypes.default.oneOfType([_propTypes.default.arrayOf(_propTypes.default.node), _propTypes.default.node])
};
var _default = SideNavMenuWithBackForward;
exports.default = _default;
