// @carbon/ibmdotcom-react/es/components/carbon-components-react/UIShell/SideNavMenu.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";
import _assertThisInitialized from "@babel/runtime/helpers/assertThisInitialized";
import _inherits from "@babel/runtime/helpers/inherits";
import _defineProperty from "@babel/runtime/helpers/defineProperty";

/* eslint-disable */

/* istanbul ignore file */

/**
 * Copyright IBM Corp. 2016, 2020
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import ChevronDown20 from '@carbon/icons-react/es/chevron--down/20';
import ddsSettings from '@carbon/ibmdotcom-utilities/es/utilities/settings/settings';
import settings from 'carbon-components/es/globals/js/settings';
import cx from 'classnames';
import PropTypes from 'prop-types';
import React from 'react';
import SideNavIcon from './SideNavIcon';
var stablePrefix = ddsSettings.stablePrefix;
var prefix = settings.prefix;
export var SideNavMenu = /*#__PURE__*/function (_React$Component) {
  _inherits(SideNavMenu, _React$Component);

  function SideNavMenu(props) {
    var _this;

    _classCallCheck(this, SideNavMenu);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(SideNavMenu).call(this, props));

    _defineProperty(_assertThisInitialized(_this), "handleToggleExpand", function (event) {
      var onToggle = _this.props.onToggle;
      event.persist();

      _this.setState(function (state) {
        return {
          isExpanded: !state.isExpanded
        };
      }, function () {
        if (onToggle) {
          onToggle(event, {
            isExpanded: _this.state.isExpanded
          });
        }
      });
    });

    _defineProperty(_assertThisInitialized(_this), "handleKeyToggleExpand", function (event) {
      if (event.charCode === 13 || event.charCode === ' ') {
        var onToggle = _this.props.onToggle;
        event.persist();

        _this.setState(function (state) {
          return {
            isExpanded: !state.isExpanded
          };
        }, function () {
          if (onToggle) {
            onToggle(event, {
              isExpanded: _this.state.isExpanded
            });
          }
        });
      }
    });

    _defineProperty(_assertThisInitialized(_this), "_renderSideNavItem", function (item, index) {
      if (item) {
        return React.cloneElement(item, {
          onClick: item.props.isbackbutton === 'true' ? _this.handleToggleExpand.bind(_assertThisInitialized(_this)) : null,
          onKeyPress: item.props.isbackbutton === 'true' ? _this.handleKeyToggleExpand.bind(_assertThisInitialized(_this)) : null
        });
      }
    });

    _this.state = {
      isExpanded: props.defaultExpanded || false,
      wasPreviouslyExpanded: props.defaultExpanded || false
    };
    return _this;
  }

  _createClass(SideNavMenu, [{
    key: "render",
    value: function render() {
      var _cx;

      var _this$props = this.props,
          buttonRef = _this$props.buttonRef,
          customClassName = _this$props.className,
          children = _this$props.children,
          IconElement = _this$props.renderIcon,
          isActive = _this$props.isActive,
          title = _this$props.title,
          large = _this$props.large,
          isbackbutton = _this$props.isbackbutton,
          rest = _objectWithoutProperties(_this$props, ["buttonRef", "className", "children", "renderIcon", "isActive", "title", "large", "isbackbutton"]);

      var isExpanded = this.state.isExpanded;
      var hasActiveChild;

      if (children) {
        // if we have children, either a single or multiple, find if it is active
        hasActiveChild = Array.isArray(children) ? children.some(function (child) {
          if (child && child.props && (child.props.isActive === true || child.props['aria-current'])) {
            return true;
          }

          return false;
        }) : children.props && (children.props.isActive === true || children.props['aria-current']);
      }

      var className = cx((_cx = {}, _defineProperty(_cx, "".concat(prefix, "--side-nav__item"), true), _defineProperty(_cx, "".concat(prefix, "--side-nav__item--active"), isActive || hasActiveChild && !isExpanded), _defineProperty(_cx, "".concat(prefix, "--side-nav__item--icon"), IconElement), _defineProperty(_cx, "".concat(prefix, "--side-nav__item--large"), large), _defineProperty(_cx, customClassName, !!customClassName), _cx));
      return React.createElement("li", {
        className: className
      }, React.createElement("button", {
        "aria-haspopup": "true",
        "aria-expanded": isExpanded,
        className: "".concat(prefix, "--side-nav__submenu"),
        onClick: this.handleToggleExpand,
        ref: buttonRef,
        type: "button"
      }, IconElement && React.createElement(SideNavIcon, null, React.createElement(IconElement, null)), React.createElement("span", {
        className: "".concat(prefix, "--side-nav__submenu-title"),
        "data-autoid": "".concat(rest.autoid)
      }, title), React.createElement(SideNavIcon, {
        className: "".concat(prefix, "--side-nav__submenu-chevron"),
        small: true
      }, React.createElement(ChevronDown20, null))), React.createElement("ul", {
        className: "".concat(prefix, "--side-nav__menu"),
        role: "menu"
      }, React.Children.map(children, this._renderSideNavItem)));
    }
  }]);

  return SideNavMenu;
}(React.Component);

_defineProperty(SideNavMenu, "propTypes", {
  /**
   * Provide an optional class to be applied to the containing node
   */
  className: PropTypes.string,

  /**
   * Provide <SideNavMenuItem>'s inside of the `SideNavMenu`
   */
  children: PropTypes.node,

  /**
   * Pass in a custom icon to render next to the `SideNavMenu` title
   */
  renderIcon: PropTypes.oneOfType([PropTypes.func, PropTypes.object]),

  /**
   * Specify whether the `SideNavMenu` is "active". `SideNavMenu` should be
   * considered active if one of its menu items are a link for the current
   * page.
   */
  isActive: PropTypes.bool,

  /**
   * Provide the text for the overall menu name
   */
  title: PropTypes.string.isRequired,

  /**
   * Specify whether the menu should default to expanded. By default, it will
   * be closed.
   */
  defaultExpanded: PropTypes.bool,

  /**
   * Property to indicate if the side nav container is open (or not). Use to
   * keep local state and styling in step with the SideNav expansion state.
   */
  isSideNavExpanded: PropTypes.bool,

  /**
   * Specify if this is a large variation of the SideNavMenu
   */
  large: PropTypes.bool,

  /**
   * For submenu back button to toggle expand/collapse
   */
  isbackbutton: PropTypes.string,

  /**
   * A callback that is called when this side nav menu is toggled by user gesture.
   */
  onToggle: PropTypes.func
});

_defineProperty(SideNavMenu, "defaultProps", {
  defaultExpanded: false,
  isActive: false,
  large: false
});

_defineProperty(SideNavMenu, "getDerivedStateFromProps", function (props, state) {
  var derivedState = null;

  if (props.isSideNavExpanded === false && state.isExpanded === true) {
    derivedState = {
      isExpanded: props.isSideNavExpanded,
      wasPreviouslyExpanded: true
    };
  } else if (props.isSideNavExpanded === true && state.wasPreviouslyExpanded === true) {
    derivedState = {
      isExpanded: props.isSideNavExpanded,
      wasPreviouslyExpanded: false
    };
  }

  return derivedState;
});

export default React.forwardRef(function (props, ref) {
  return React.createElement(SideNavMenu, _extends({}, props, {
    buttonRef: ref
  }));
});
